ANDROID: Add real 3D support to Android platform

lephilousophe opened the pull request:
This is a first stab to add 3D support through GLES2 to the Android platform.
The patch is based on android3d backend which is now unmaintained and on SDL backend for the graphics backend switching.

sev- commented on backends/graphics3d/android/texture.h:
```diff
@@ -0,0 +1,312 @@
+#ifndef BACKENDS_GRAPHICS3D_ANDROID_TEXTURE_H
+#define BACKENDS_GRAPHICS3D_ANDROID_TEXTURE_H
+
+#if defined(__ANDROID__)
```
Another check...

lephilousophe replied:
Done, leftover from android3d platform import.

sev- commented on backends/platform/android/events.cpp:
```diff
@@ -373,6 +373,13 @@ static const Common::KeyCode jkeymap[] = {
 
+void OSystem_Android::updateEventScale(uint32 w, uint32 h) {
+	if ((h != 0) && (w != 0)) {
+	    _eventScaleY = 100 * 480 / h;
```
What are these magic numbers?

lephilousophe replied:
This a bare copy from android3d backend.
In current code, the `_eventScaleX` and `_eventScaleY` are set to 100 and never touched again.
android3d has this magical function. I suppose 480 is related to good ol' 640x480 and 100 is to have a percentage but I don't know the rationale behind this.
This scale seems to be used by directional pads, trackballs and joysticks and I don't have any to test on my android.

aquadran replied:
maybe add comments in code for now

lephilousophe replied:
After some digging in codebase, this function did exist in the past and got removed by #1695.
I don't know how useful it was. Maybe @ccawley2011 knows better.
Removing for now. We can add it back later if we want to.

aquadran replied:
indeed

sev- commented on backends/platform/android/touchcontrols.cpp:
```diff
@@ -0,0 +1,362 @@
+ *
+ */
+
+#if defined(__ANDROID__)
```
Another unnecesary check

lephilousophe replied:
Done, leftover from android3d platform import

sev- commented on backends/platform/android/touchcontrols.cpp:
```diff
@@ -0,0 +1,362 @@
+}
+
+TouchControls::Pointer *TouchControls::getPointerFromId(int ptrId, bool createNotFound) {
+	unsigned int freeEntry = -1;
```
I'd recommend using `uint` here and below for brevity

lephilousophe replied:
Done.

sev- commented on backends/platform/android/touchcontrols.h:
```diff
@@ -0,0 +1,112 @@
+#ifndef ANDROID_TOUCHCONTROLS_H_
+#define ANDROID_TOUCHCONTROLS_H_
+
+#if defined(__ANDROID__)
```
Not needed?

lephilousophe replied:
Done, leftover from android3d platform import

sev- reviewed (approved):
Excellent work!

I had some minor notes.

I was not looking too deep into implementation, but it seems that the backend complexity grew a lot with this. I wonder, how easy it would be to implement new backends or add 3d support to iOS....

aquadran commented:
I think 2D 

> Excellent work!
> 
> I had some minor notes.
> 
> I was not looking too deep into implementation, but it seems that the backend complexity grew a lot with this. I wonder, how easy it would be to implement new backends or add 3d support to iOS....

this is done same way as for SDL. for long term both should be merged. so for iOS, that depends which approach is taken, separated backend or changes integrated into one backend.

aquadran commented on backends/graphics3d/android/texture.cpp:
```diff
@@ -0,0 +1,469 @@
+	delete[] _pixels;
+
+	_pixels = new byte[w * h];
+	assert(_pixels);
```
handle allocation error

aquadran commented on backends/graphics3d/android/texture.cpp:
```diff
@@ -0,0 +1,469 @@
+	_palettePixelFormat = pixelFormat;
+	_fake_format = Graphics::PixelFormat::createFormatCLUT8();
+
+	_palette = new uint16[256];
```
handle allocation fail

aquadran commented on backends/platform/android/touchcontrols.cpp:
```diff
@@ -0,0 +1,362 @@
+	uint16 w = dec.getSurface()->w;
+	uint16 h = dec.getSurface()->h;
+	uint16 pitch = dec.getSurface()->pitch;
+	ret->allocBuffer(w, h);
```
any fail handling here?

aquadran commented on backends/platform/android/touchcontrols.h:
```diff
@@ -0,0 +1,112 @@
+
+	struct Pointer {
+		Pointer() : id(-1), startX(-1), startY(-1), currentX(-1), currentY(-1), function(kFunctionNone), active(false) {}
+		void reset() { id = -1; startX = startY = currentX = currentY = -1; function = kFunctionNone; active = false; }
```
formatting

lephilousophe replied:
Fixed using astyle.

aquadran commented on backends/platform/android/touchcontrols.cpp:
```diff
@@ -0,0 +1,362 @@
+#include "backends/platform/android/touchcontrols.h"
+
+TouchControls::TouchControls() :
+	_arrows_texture(NULL),
```
can be nullptr

aquadran commented on backends/platform/android/jni-android.cpp:
```diff
@@ -713,6 +730,18 @@ void JNI::pushEvent(JNIEnv *env, jobject self, int type, int arg1, int arg2,
+
+	assert(_system);
+
+	_system->getTouchControls().update((TouchControls::Action) action, ptr, x, y);
```
maybe condition for _system. not sure if it's possible to happen

aquadran commented on backends/graphics3d/android/texture.cpp:
```diff
@@ -0,0 +1,469 @@
+}
+
+void GLESFakePaletteTexture::fillBuffer(uint32 color) {
+	assert(_surface.getPixels());
```
add condition for null surface just in case

aquadran commented on backends/graphics3d/android/texture.cpp:
```diff
@@ -0,0 +1,469 @@
+
+	fillBuffer(0);
+
+	_buf = new uint16[w * h];
```
handle allocation error

aquadran commented on backends/graphics3d/android/texture.cpp:
```diff
@@ -0,0 +1,469 @@
+	setDirtyRect(Common::Rect(x, y, x + w, y + h));
+
+	const byte *src = (const byte *)buf;
+	byte *dst = _pixels + y * _surface.pitch + x;
```
handle allocation error

aquadran commented on backends/graphics3d/android/texture.cpp:
```diff
@@ -0,0 +1,469 @@
+}
+
+void GLESTexture::fillBuffer(uint32 color) {
+	assert(_surface.getPixels());
```
handle null condition

aquadran commented on backends/graphics3d/android/android-graphics3d.cpp:
```diff
@@ -0,0 +1,982 @@
+#ifdef USE_RGB_COLOR
+	initTexture(&_game_texture, width, height, format);
+#else
+	_game_texture->allocBuffer(width, height);
```
handle allocation error

aquadran commented on backends/graphics3d/android/android-graphics3d.cpp:
```diff
@@ -0,0 +1,982 @@
+	// setMouseCursor.  We need the palette allocated before
+	// setMouseCursor however, so just take a guess at the desired
+	// size (it's small).
+	_mouse_texture_palette->allocBuffer(20, 20);
```
handle allocation error

aquadran commented on backends/graphics3d/android/android-graphics3d.cpp:
```diff
@@ -0,0 +1,982 @@
+	initViewport();
+
+	// resize game texture
+	initSizeIntern(width, height, 0);
```
handle error

aquadran commented on backends/graphics3d/android/android-graphics3d.cpp:
```diff
@@ -0,0 +1,982 @@
+	GLTHREADCHECK;
+
+	Graphics::Surface *surface = _game_texture->surface();
+	assert(surface->getPixels());
```
handle error

lephilousophe commented on backends/graphics3d/android/android-graphics3d.cpp:
```diff
@@ -0,0 +1,982 @@
+ *
+ */
+
+#if defined(__ANDROID__)
```
Done, leftover from android3d platform import

lephilousophe commented on backends/graphics3d/android/android-graphics3d.cpp:
```diff
@@ -0,0 +1,982 @@
+#else
+	_game_texture->allocBuffer(width, height);
+#endif
+#ifdef USE_GLES2
```
Done, leftover from android3d platform import

lephilousophe commented on backends/graphics3d/android/texture.cpp:
```diff
@@ -0,0 +1,469 @@
+ *
+ */
+
+#if defined(__ANDROID__)
```
Done, leftover from android3d platform import

lephilousophe commented on backends/graphics3d/android/android-graphics3d.cpp:
```diff
@@ -0,0 +1,982 @@
+	_mouse_keycolor(0),
+	_show_mouse(false),
+	_use_mouse_palette(false)
+{
```
Done, leftover from android3d platform import. Fixed using astyle.

lephilousophe commented on backends/graphics3d/android/android-graphics3d.cpp:
```diff
@@ -0,0 +1,982 @@
+					if (*s++ == (keycolor & 0xffff))
+						*d = 0;
+		}
+		else if (format->bytesPerPixel == 4) {
```
Done, leftover from android3d platform import. Fixed using astyle.

lephilousophe commented on backends/graphics3d/android/texture.cpp:
```diff
@@ -0,0 +1,469 @@
+// Supported GL extensions
+static bool npot_supported = false;
+
+OpenGL::ShaderGL * g_box_shader;
```
Done, leftover from android3d platform import. Fixed using astyle.

lephilousophe commented on backends/graphics3d/android/texture.cpp:
```diff
@@ -0,0 +1,469 @@
+	_pixelFormat(pixelFormat),
+	_palettePixelFormat(),
+	_is_game_texture(false)
+{
```
Done, leftover from android3d platform import. Fixed using astyle.

lephilousophe commented on backends/graphics3d/android/texture.cpp:
```diff
@@ -0,0 +1,469 @@
+	_palette(0),
+	_pixels(0),
+	_buf(0)
+{
```
Done, leftover from android3d platform import. Fixed using astyle.

lephilousophe commented on backends/graphics3d/android/android-graphics3d.cpp:
```diff
@@ -0,0 +1,982 @@
+		setCursorPaletteInternal(colors, start, num);
+
+	const Graphics::PixelFormat &pf = _game_texture->getPalettePixelFormat();
+	byte *p = _game_texture->palette() + start * 2;
```
The format is always 16bpp because the object is a GLESFakePalette565Texture.
I added a comment and used sizeof(uint16) instead.

aquadran replied:
can we do some assert here if someone change texture format? like assert(pf.bpp() == 16);

lephilousophe replied:
done

lephilousophe commented on backends/graphics3d/android/android-graphics3d.cpp:
```diff
@@ -0,0 +1,982 @@
+	if (_mouse_texture == _mouse_texture_palette) {
+		assert(keycolor < 256);
+
+		byte *p = _mouse_texture_palette->palette() + _mouse_keycolor * 2;
```
The format is always 16bpp because the object is a GLESFakePalette5551Texture.
I added a comment and used sizeof(uint16) instead.

lephilousophe commented on backends/graphics3d/android/android-graphics3d.cpp:
```diff
@@ -0,0 +1,982 @@
+		uint8 r, g, b;
+
+		for (uint i = 0; i < 256; ++i, src += 2, dst += 2) {
+			pf_src.colorToRGB(READ_UINT16(src), r, g, b);
```
same as above.

lephilousophe commented on backends/graphics3d/android/android-graphics3d.cpp:
```diff
@@ -0,0 +1,982 @@
+	int overlay_width = MAX(JNI::egl_surface_width, 320);
+	int overlay_height = MAX(JNI::egl_surface_height, 200);
+
+	// the 'normal' theme layout uses a max height of 400 pixels. if the
```
It was already commented in android3d. Removed.

lephilousophe commented on backends/graphics3d/android/texture.cpp:
```diff
@@ -0,0 +1,469 @@
+#include "backends/graphics3d/android/texture.h"
+
+// Supported GL extensions
+static bool npot_supported = false;
```
It is now.

aquadran commented on backends/graphics/android/android-graphics.cpp:
```diff
@@ -143,3 +142,32 @@ bool AndroidGraphicsManager::notifyMousePosition(Common::Point &mouse) {
+	state.aspectRatio   = getFeatureState(OSystem::kFeatureAspectRatioCorrection);
+	state.fullscreen    = getFeatureState(OSystem::kFeatureFullscreenMode);
+	state.cursorPalette = getFeatureState(OSystem::kFeatureCursorPalette);
+#ifdef USE_RGB_COLOR
```
do we need "#ifdef USE_RGB_COLOR" for android backend?

lephilousophe commented on backends/graphics3d/android/texture.cpp:
```diff
@@ -0,0 +1,469 @@
+GLESFakePaletteTexture::GLESFakePaletteTexture(GLenum glFormat, GLenum glType,
+									Graphics::PixelFormat pixelFormat) :
+	GLESBaseTexture(glFormat, glType, pixelFormat),
+	_palette(0),
```
Done with a few other places.

lephilousophe commented on backends/graphics3d/android/android-graphics3d.cpp:
```diff
@@ -0,0 +1,982 @@
+
+	GLTHREADCHECK;
+
+#ifdef USE_RGB_COLOR
```
I think it was here in android3d because the backend could be used with 2D games.
In this PR this graphical backend is only for 3D so I think it could be removed.
It depends on if we want to reuse this code later to merge both graphical backends.

aquadran replied:
I think it was copy and paste from 2d backend

lephilousophe replied:
On a second thought, it is needed if someone wants to build with configure option `--disable-16bit`. In this case the `USE_RGB_COLOR` is undefined and some functions are unavailable.

aquadran replied:
so it's rather build thing, not runtime, since no such case

lephilousophe replied:
Yes indeed

lephilousophe commented on backends/graphics3d/android/texture.cpp:
```diff
@@ -0,0 +1,469 @@
+	delete[] _pixels;
+
+	_pixels = new byte[w * h * _surface.format.bytesPerPixel];
+	assert(_pixels);
```
The `new` operator will either throw an exception if libstdc++ is compiled with exceptions support or will abort.
This assert is useless and has been removed. Same for others.
If we can't allocate memory, I don't know which graceful action we could do.

sev- commented:
How close is this to be mergeable?

lephilousophe commented:
@aquadran said that GLAD PR should be merged before in [this comment ](https://github.com/scummvm/scummvm/pull/3463#issuecomment-957269416).

aquadran commented:
since GLAD depends on release branch. that make no sense to wait for it. I was rather about PR collisions. but in this case I think we should not wait, since we need to test it on master.

lephilousophe commented:
OK, I will rebase this next week-end at worst.

lephilousophe commented:
I just rebased against current master, fixed conflicts and fixed the last opened review.
What misses IMO is to use a keymapper for the virtual controls.
For now it hardcodes the `KEY_` constants. My main problem is that the virtual controls are only enabled when the 3D game is run. This means that we can't configure the keymapper with events from these.
On the other hand, I didn't look at integrating the virtual controls on 2D side. It could be useful, as it lets the user have more controls than just touch, back key and virtual keyboard.

lephilousophe commented:
Updated to make it not use PixelBuffer. Following the code cleanup, `game_pbuf` was never used anyway so it was easy to remove.

bluegr commented:
Awesome work! Well done

lephilousophe commented:
Just changed the virtual controls to issue joystick buttons events instead of hard coded keycodes.
It's not yet ideal because we can't customize them in the UI but I think it's more intuitive.
